### PR TITLE
Update the Date data set to avoid DST transition windows

### PR DESCRIPTION
**Pull request by logiclrd**
This PR:

* Adds a unit test that I'm _pretty sure_ will find a DST window and force a value to generate inside of it. Works on my system, anyway. The test should automatically skip if the local time zone doesn't do DST transitions.
* Updates the implementation of the `Date` data set so that all requests get funnelled into `Between` and `BetweenOffset`, and then reimplements `Between` and `BetweenOffset` to convert the range to UTC first, then obtain a random value, and then convert it back to a local time if appropriate before returning it, taking advantage of the framework's daylight savings transition calculations.

The unit test fails with the original `Date` data set code, and the change in the `Date` data set fixes the unit test. On my system, at least.

This PR is in response to #365.

### CODE REVIEW

**Comment by logiclrd** [comment]
Rebased. I thought I was working with the latest `master`, but I wasn't. The change is now in terms of the latest `master`.

File: Source/Bogus/DataSets/Date.cs [hunk 1]
```diff
@@ -161,14 +153,19 @@ public DateTime Between(DateTime start, DateTime end)
...
+         var maxTicks = Math.Max(startTicks, endTicks);
 
          var totalTimeSpanTicks = maxTicks - minTicks;
 
          var partTimeSpan = RandomTimeSpanFromTicks(totalTimeSpanTicks);
 
-         return new DateTimeOffset(minTicks, start.Offset) + partTimeSpan;
+         var dateTime = new DateTime(minTicks, DateTimeKind.Unspecified) + partTimeSpan;
+
+         return new DateTimeOffset(dateTime + start.Offset, start.Offset);
```
Comment (bchavez) [inline Source/Bogus/DataSets/Date.cs:168]: I'm going to replace these two lines with:

```csharp
         if( start < end )
            return start + partTimeSpan;
         else
            return end + partTimeSpan;
```

Given an invalid `start` time; the function will still return an invalid `DateTimeOffset`.

image

With the code change, at least the method returns a valid time, albeit outside the "specified" time range; which is fine I guess, because as it stands now, it would still have been outside the "2:59 AM - 3:00 AM" time as noted above.

image

Comment (logiclrd) [reply Source/Bogus/DataSets/Date.cs:168]: That seems legit :-)

Comment (logiclrd) [reply Source/Bogus/DataSets/Date.cs:168]: I guess to my mind, if the caller supplies impossible DateTime values to begin with, all bets are off. My specific use case doesn't intersect with that, so I don't have strong opinions about how exactly it behaves in that edge case.

Comment (logiclrd) [reply Source/Bogus/DataSets/Date.cs:168]: Thinking it through further, if one wanted to be _really_ thorough, then the algorithm would need to include some sort of "compute _real_ DateTime range" functionality, so that `start` and `end` were adjusted if they lay inside an impossible time span, and this would then also throw if `start` and `end` both lay within the same transition window (so that no real date/time `x` would satisfy `start <= x <= end`). This would be in addition to generating the actual bogus value in UTC, to avoid transition windows in the middle of the range. I think this would be the most polished form of this, but I'm not sure if it's worth the effort.

Comment (logiclrd) [reply Source/Bogus/DataSets/Date.cs:168]: Pseudocode:
```
void ComputeRealRange(ref DateTime start, ref DateTime end)
{
  if (start > end)
    Swap(ref start, ref end);

  var (windowStart, windowEnd) = GetDSTTransitionWindow(start.Year);

  if (windowStart <= start <= windowEnd)
    start = windowEnd;

  (windowStart, windowEnd) = GetDSTTransitionWindow(end.Year);

  if (windowStart <= end <= windowEnd)
    end = windowStart;

  if (start > end)
    throw new Exception("Impossible");
}
```

Comment (logiclrd) [reply Source/Bogus/DataSets/Date.cs:168]: With this approach, a supplied range as shown where `start` lay within the impossible range and `end` was exactly the first next valid `DateTime` value, it would always return exactly the `end` value because the _effective_ `start` would be bumped to the end of the transition window and would coincide with `end`.

**Comment by logiclrd** [comment]
Any changes you'd like to request on this PR? :-)

**Comment by logiclrd** [comment]
I have implemented the proposed algorithm for excluding impossible values from the range by bumping `start` forward and `end` back, along with unit tests that detect the correction.

**Comment by logiclrd** [comment]
There were some shenanigans surrounding `Kind` and offsets but that's now sorted out. For your test, with `start` set to 1 minute before the end of the transition and `end` set to exactly the end of the transition, the requested range actually only has one real `DateTime` value, that being the exact end of the transition, and so it will always return that exact value.
```
2021-03-14 2:59:00 AM -06:00
2021-03-14 3:00:00 AM -05:00
2021-03-14 8:59:00 AM +00:00
2021-03-14 8:00:00 AM +00:00
2021-03-14 3:00:00 AM -06:00
```

**Comment by logiclrd** [comment]
Sorry, I don't normally do so many force-pushes, but in my eagerness I'm afraid I pushed changes only to realize a few minutes later that I had overlooked a case and return to the code to address it.

**Comment by bchavez** [comment]
@logiclrd no worries. Thank you for making the improvements; sorry for the delay getting this merged. I've been busy with a few life events that are taking a bit of time to resolve.

File: Source/Bogus.Tests/DataSetTests/DateTest.cs [hunk 2]
```diff
@@ -352,5 +353,236 @@ public void can_get_timezone_string()
       {
          date.TimeZoneString().Should().Be("Asia/Yerevan");
       }
+
+      public class FactWhenDaylightSavingsSupported : FactAttribute
+      {
+         public FactWhenDaylightSavingsSupported()
+         {
+            if (!TimeZoneInfo.Local.SupportsDaylightSavingTime)
```
Comment (Chobocharles) [inline Source/Bogus.Tests/DataSetTests/DateTest.cs:361]: Wow, I had no clue this was a thing.

File: Source/Bogus.Tests/DataSetTests/DateTest.cs [hunk 3]
```diff
@@ -352,5 +353,236 @@ public void can_get_timezone_string()
...
+
+         faker.Random = new Randomizer(localSeed: 5);
+
+         var dstRules = TimeZoneInfo.Local.GetAdjustmentRules();
+
+         var now = DateTime.Now;
+
+         var effectiveRule = dstRules.Single(rule => (rule.DateStart <= now) && (rule.DateEnd >= now));
+
+         var transitionStartTime = CalculateTransitionDateTime(now, effectiveRule.DaylightTransitionStart);
```
Comment (Chobocharles) [inline Source/Bogus.Tests/DataSetTests/DateTest.cs:382]: New to this code base, but what's the consensus on generating test helper methods or classes that can contain boiler plate construction and arrangement of objects?

File: Source/Bogus.Tests/DataSetTests/DateTest.cs [hunk 4]
```diff
@@ -352,5 +353,236 @@ public void can_get_timezone_string()
...
```
Comment (Chobocharles) [inline Source/Bogus.Tests/DataSetTests/DateTest.cs:411]: Because we can probably toss all of this initial setup into a base method (except for the rule) and just call the method at the beginning of the tests

File: Source/Bogus.Tests/DataSetTests/DateTest.cs (hunk 3 above)
Comment (logiclrd) [reply Source/Bogus.Tests/DataSetTests/DateTest.cs:382]: My convention when writing tests has been to make the test methods completely logically independent of one another, but I have seen tests and test suites that reuse a great deal of setup. This could be refactored, but was written this way intentionally (if without a great deal of thought, per se).

**Comment by logiclrd** [comment]
@bchavez Ping :-)

**Comment by bchavez** [comment]
Hey @logiclrd , im still here :) I'll try to give it a shot this weekend!!

**Comment by logiclrd** [comment]
@bchavez How's it going? :-D

**Comment by bchavez** [comment]
hey @logiclrd , for sure, I need to get back to this - I just haven't found enough time to do the extensive testing I want to do before I merge this PR

**Comment by bchavez** [comment]
@logiclrd , it was the same test I did that found the previous issue in the PR here:
* https://github.com/bchavez/Bogus/pull/366#discussion_r598342181

I know it is probably fixed, but I need to see it first-hand and probably see how it behaves in other scenarios too.

**Comment by logiclrd** [comment]
The code has changed since then. I'll run the same test, see what happens. :-)

**Comment by logiclrd** [comment]
Okay, I added a test that I think encapsulates what you were talking about, and I tweaked the behaviour of the `Date` set to match the expectations encoded in the tests.

**Comment by logiclrd** [comment]
Let me know if that test covers the functionality you were thinking of. :-)

**Comment by bchavez** [comment]
Thank you @logiclrd -- I appreciate it; hopefully, after I get personal and business taxes out of the way I can take a 2nd look. Just really difficult to do OSS work after a full-time job; and I get pretty tired by the end of the day.

There's always a risk; because as soon as I merge and release this code, I have to be ready for an influx of support requests which is why I'm waiting until I have enough buffer-time to handle any immediate issue that could come up as a result of some **_new_** time code bug that could come up as a result of this new code.

Yes, we do have a bug now, but at least it's known -- and hasn't caused any major problems as far as GH issues are concerned.

**Comment by logiclrd** [comment]
That's fair :-) For what it's worth, we've been running on a private prerelease build on our private Azure Artifacts server for, well, for some time now :-P

**Comment by logiclrd** [comment]
How's it going? :-)

**Comment by logiclrd** [comment]
Where are we at?

**Comment by voronoipotato** [comment]
Just hit this trying to test an issue I was having where the unit test was working locally but failing on the build server configured to use UTC. 
```F#
      Expected: ···2022-03-29 05:46:04Z
      Actual:   ···2022-03-29 05:46:03Z 
```

building this test was actually quite difficult, because the faker facade really really doesn't want to let go of the seed even though I created a whole new instance. Turns out you have to set the seed on the original faker instance and not the new one, but you also do need to create a new instance, and you need to do it before you update the seed. 

```F#
System.Threading.Thread.CurrentThread.CurrentCulture <- CultureInfo.InvariantCulture
let faker =Faker(Random = (Randomizer(777), Localization = "en")
let data = generateDates faker //some code to make a ton of dates
let faker2 =Faker(Random = (Randomizer(777), Localization = "en")
faker.Random <- Randomizer(777) 
//if I don't do this for some reason faker produces entirely different datasets even though faker2 should be an entirely different instance, also setting faker.Random on faker 2 does nothing at all. 
let data2 = generateDates faker2
Assert.IsTrue(data.ToString() = data2.ToString())
```

probably going to create a new ticket for this since it can make it nearly impossible to replicate bugs without determinism, and it would be nice if separate faker instances were at least "able" to be completely separate.

**Comment by logiclrd** [comment]
@bchavez Any chance of getting this loose end tied off? :-)

**Comment by logiclrd** [comment]
Oh, you know what, I was mistaken. We don't have a custom build of Bogus. We have a custom build of FluentAssertions with support for System.Data types. We just still get random test failures around DST transition. Another developer at some point made it retry tests up to 5 times to handle stochastic failures. That's its own thing :-P

**Comment by logiclrd** [comment]
Maybe we can get this in place before the next DST switch? :-)

**Comment by logiclrd** [comment]
The next DST switch is in 5 days.

**Comment by logiclrd** [comment]
Well, I guess the next DST switch _that matters_ is March 12, 2023. :-P

**Comment by logiclrd** [comment]
@bchavez Ping :-)

**Comment by logiclrd** [comment]
10 days away!

**Comment by logiclrd** [comment]
Hello from the other side :-)

**Comment by bchavez** [comment]
Hey @ChristoWolf, if you have a chance or any free time;

Could you double-check this PR and see if we have sufficient unit test coverage for edge cases? I'd like to know your thoughts on this PR. 

I'm curious if there are any edge cases where this new code might break existing code. Or ways we could intentionally break the new code or its assumptions.

Right now, currently working on fixing the CI/CD build system and getting the pipeline up and running again. After that I'll likely switch gears to this PR; but would be nice to have a fresh set of eyes and opinions on it.

**Comment by logiclrd** [comment]
I'm not saying that it's impossible that there might yet be bugs, but I don't _think_ they are, and the _very purpose_ of the reworked code is to completely rule out edge cases with impossible dates by design. If you can find a way to break assumptions with the new code, though, that'll be good :-) Better to find it out now!

**Comment by ChristoWolf** [comment]
@bchavez: Sure, I'd love to, hope I can do it already today!

**Comment by ChristoWolf** [comment]
@bchavez: Not sure why, but I am having big troubles with building the solution, fake is messing up no matter which version I use.
I don't think I had this issue in the past.
In any case, I will for now review the PR without running the changes.

**Comment by bchavez** [comment]
@ChristoWolf ah, for sure, the F# FAKE stuff is busted right now; I think the root cause is mostly in `\Source\Builder\build.fsx.lock` files that attempt to get resolved, but are fundamentally not compatible with the latest .NET 7 SDK.

As a workaround, I think you might be able to rely on the base CLI commands like `dotnet restore; dotnet build; dotnet test` in the `\Source` folder. Also, I think opening the SLN and building/testing should work from within the IDE itself too.

Let me know if there's anything else I can try to help with.

**Comment by ChristoWolf** [comment]
@bchavez: Good to know that it's not just me :)
Got it working, thanks!

File: Source/Bogus/DataSets/Date.cs [hunk 5]
```diff
@@ -144,31 +129,221 @@ public DateTimeOffset FutureOffset(int yearsToGoForward = 1, DateTimeOffset? ref
       /// <param name="end">End time</param>
       public DateTime Between(DateTime start, DateTime end)
       {
-         var minTicks = Math.Min(start.Ticks, end.Ticks);
-         var maxTicks = Math.Max(start.Ticks, end.Ticks);
+         ComputeRealRange(ref start, ref end, start.Kind, out var preferRangeBoundary);
+
+         var startTicks = start.ToUniversalTime().Ticks;
```
Comment (ChristoWolf) [inline Source/Bogus/DataSets/Date.cs:134]: I don't know all the ins and outs, but I think it is still [possible to hit problematic transitions using UTC](https://stackoverflow.com/a/40626617).
Could a HPET help with this?

File: Source/Bogus/DataSets/Date.cs [hunk 6]
```diff
@@ -144,31 +129,221 @@ public DateTimeOffset FutureOffset(int yearsToGoForward = 1, DateTimeOffset? ref
...
+         if (transition.IsFixedDateRule)
+         {
+            startTime = new DateTime(
+               dateTime.Year,
+               transition.Month,
+               transition.Day,
+               transition.TimeOfDay.Hour,
+               transition.TimeOfDay.Minute,
+               transition.TimeOfDay.Second,
+               transition.TimeOfDay.Millisecond);
```
Comment (ChristoWolf) [inline Source/Bogus/DataSets/Date.cs:270]: I guess you could reduce nesting by already returning in here.

File: Source/Bogus/DataSets/Date.cs [hunk 7]
```diff
@@ -144,31 +129,221 @@ public DateTimeOffset FutureOffset(int yearsToGoForward = 1, DateTimeOffset? ref
...
+         }
+         else
+         {
+            var calendar = CultureInfo.CurrentCulture.Calendar;
+
+            var startOfWeek = transition.Week * 7 - 6;
```
Comment (ChristoWolf) [inline Source/Bogus/DataSets/Date.cs:276]: Not sure if such `Calendar`s are implemented, but there are real-world calendars which do not follow a 7-day week.

File: Source/Bogus/DataSets/Date.cs [hunk 8]
```diff
@@ -144,31 +129,221 @@ public DateTimeOffset FutureOffset(int yearsToGoForward = 1, DateTimeOffset? ref
...
+
+            var firstDayOfWeek = (int)calendar.GetDayOfWeek(new DateTime(dateTime.Year, transition.Month, 1));
+            var changeDayOfWeek = (int)transition.DayOfWeek;
+
+            int transitionDay =
```
Comment (ChristoWolf) [inline Source/Bogus/DataSets/Date.cs:281]: I guess this could be `var`.

File: Source/Bogus/DataSets/Date.cs (hunk 7 above)
Comment (ChristoWolf) [reply Source/Bogus/DataSets/Date.cs:276]: The same is relevant for a few more lines in this method.

File: Source/Bogus/DataSets/Date.cs [hunk 9]
```diff
@@ -144,31 +129,221 @@ public DateTimeOffset FutureOffset(int yearsToGoForward = 1, DateTimeOffset? ref
...
       {
-         var minTicks = Math.Min(start.Ticks, end.Ticks);
-         var maxTicks = Math.Max(start.Ticks, end.Ticks);
+         var startTime = start.ToUniversalTime().DateTime;
+         var endTime = end.ToUniversalTime().DateTime;
 
-         var totalTimeSpanTicks = maxTicks - minTicks;
+         if (startTime > endTime)
+            return end;
+         else
```
Comment (ChristoWolf) [inline Source/Bogus/DataSets/Date.cs:334]: Redundant `else` due to the `return` in the `if` block.

File: Source/Bogus.Tests/DataSetTests/DateTest.cs [hunk 10]
```diff
@@ -1,16 +1,22 @@
...
 using System.Globalization;
+using System.Linq;
 using Bogus.DataSets;
 using FluentAssertions;
 using Xunit;
+using Xunit.Abstractions;
 
 namespace Bogus.Tests.DataSetTests
 {
    public class DateTest : SeededTest
```
Comment (ChristoWolf) [inline Source/Bogus.Tests/DataSetTests/DateTest.cs:11]: Just an idea: We could also add a PInvokation to [SetTimeZoneInformation](https://learn.microsoft.com/en-us/windows/win32/api/timezoneapi/nf-timezoneapi-settimezoneinformation) (on Windows) to change the runner's time zone which would allow for covering all time zones e.g. in a  data-driven manner.
This affects system settings though, so I am not sure how compatible this is with the CI runners.

File: Source/Bogus.Tests/DataSetTests/DateTest.cs (hunk 3 above)
Comment (ChristoWolf) [reply Source/Bogus.Tests/DataSetTests/DateTest.cs:387]: I would also prefer consolidation of the test initialization, what's your opinion @bchavez?

File: Source/Bogus.Tests/DataSetTests/DateTest.cs [hunk 11]
```diff
@@ -352,5 +358,267 @@ public void can_get_timezone_string()
...
+
+         var transitionStartTime = CalculateTransitionDateTime(now, effectiveRule.DaylightTransitionStart);
+
+         // When converting back, .NET picks the end of the transition window instead of the start.
+         var transitionEndTime = transitionStartTime.ToUniversalTime().ToLocalTime();
+
+         // Act
+         var value = faker.Date.Between(transitionStartTime.AddHours(-1), transitionEndTime.AddHours(+2));
+
+         // Assert
```
Comment (ChristoWolf) [inline Source/Bogus.Tests/DataSetTests/DateTest.cs:395]: To prevent defect masking, these could be wrapped by an `AssertionScope`.

Comment (ChristoWolf) [reply Source/Bogus.Tests/DataSetTests/DateTest.cs:395]: The same also holds true for other assertion groups in your changes.

File: Source/Bogus.Tests/DataSetTests/DateTest.cs [hunk 12]
```diff
@@ -352,5 +358,267 @@ public void can_get_timezone_string()
...
+         {
+            var sample = faker.Date.Between(windowStart, windowEnd);
+
+            sample.Should().BeOnOrAfter(transitionEndTime);
+            sample.Should().BeOnOrBefore(windowEnd);
+
+            haveSampleThatIsNotWindowEnd = (sample < windowEnd);
+         }
+
+         haveSampleThatIsNotWindowEnd.Should().BeTrue(because: "the effective range should include values other than windowEnd");
```
Comment (ChristoWolf) [inline Source/Bogus.Tests/DataSetTests/DateTest.cs:435]: I guess this is really nitpicky, so feel free to ignore this:
`windowEnd` in the string would not need to be hard-coded if using `nameof(windowEnd)` instead.

File: Source/Bogus.Tests/DataSetTests/DateTest.cs [hunk 13]
```diff
@@ -352,5 +358,267 @@ public void can_get_timezone_string()
...
+               transition.TimeOfDay.Hour,
+               transition.TimeOfDay.Minute,
+               transition.TimeOfDay.Second,
+               transition.TimeOfDay.Millisecond,
+               DateTimeKind.Local);
+         }
+
+         var calendar = CultureInfo.CurrentCulture.Calendar;
+
+         var startOfWeek = transition.Week * 7 - 6;
```
Comment (ChristoWolf) [inline Source/Bogus.Tests/DataSetTests/DateTest.cs:600]: Again, I still have the question here if `7` is always correct for week-lengths.

File: Source/Bogus/DataSets/Date.cs [hunk 14]
```diff
@@ -144,31 +129,221 @@ public DateTimeOffset FutureOffset(int yearsToGoForward = 1, DateTimeOffset? ref
...
          var partTimeSpan = RandomTimeSpanFromTicks(totalTimeSpanTicks);
 
-         return new DateTime(minTicks, start.Kind) + partTimeSpan;
+         var value = new DateTime(minTicks, DateTimeKind.Utc) + partTimeSpan;
+
+         if (start.Kind != DateTimeKind.Utc)
+         {
+            value = value.ToLocalTime();
+
+            if (value > end)
```
[the remark this answers is not in the document]
Comment (logiclrd) [reply Source/Bogus/DataSets/Date.cs:157]: `DateTime` relative comparisons do not factor in the kind, so going from universal to local can change the result of `value > end`. Consider:
```
PS C:\code\Bogus\Source> $a = [DateTime]::new(2023, 3, 19, 11, 0, 0, 0, "Utc")
PS C:\code\Bogus\Source> $b = [DateTime]::new(2023, 3, 19, 11, 0, 0, 0, "Local")
PS C:\code\Bogus\Source> $c = [DateTime]::new(2023, 3, 19, 10, 30, 0, 0, "Local")
PS C:\code\Bogus\Source> $c -lt $b
True
PS C:\code\Bogus\Source> $c -lt $a
True
PS C:\code\Bogus\Source> $c -lt $a.ToLocalTime()
False
PS C:\code\Bogus\Source>
```

[the remark this answers is not in the document]
Comment (ChristoWolf) [reply Source/Bogus/DataSets/Date.cs:157]: Didn't know that, thanks for the insight!

File: Source/Bogus/DataSets/Date.cs [hunk 15]
```diff
@@ -144,31 +129,221 @@ public DateTimeOffset FutureOffset(int yearsToGoForward = 1, DateTimeOffset? ref
...
+            {
+               start = new DateTime(window.Start.Ticks, start.Kind);
+
+               if (start == end)
+                  return;
+            }
+
+            window = GetForwardDSTTransitionWindow(end);
+
+            if ((end >= window.Start) && (end < window.End))
```
[the remark this answers is not in the document]
Comment (logiclrd) [reply Source/Bogus/DataSets/Date.cs:212]: The DST transition window is the span of times that, in the local time zone, _don't exist_ because of the abrupt change do DST. But, `DateTime` comparisons don't take `Kind` into account and just look at the ticks. So, you can create `DateTime` values that point at the start and the end of the window that are _actually the same date/time_, but are numerically different. If you `.ToUtc()` them, you get the same UTC date/time out. This comparison is checking whether the desired end of the range from which to get a bogus DateTime lies within that nebulous region of date/time values that don't actually exist in the real world.

File: Source/Bogus/DataSets/Date.cs [hunk 16]
```diff
@@ -144,31 +129,221 @@ public DateTimeOffset FutureOffset(int yearsToGoForward = 1, DateTimeOffset? ref
...
+               end = new DateTime(window.End.Ticks, end.Kind);
+
+               // We had to bump the end, meaning that the end was not already a valid
+               // DateTime value (within the DST transition range), so prefer the start
+               // instead.
+               preferRangeBoundary = start;
+            }
+
+            if (start > end)
+               throw new Exception("DateTime range does not contain any real DateTime values due to daylight savings transitions");
```
[the remark this answers is not in the document]
Comment (logiclrd) [reply Source/Bogus/DataSets/Date.cs:223]: If one is using `Faker` to generate `DateTime` values, it will already be avoiding DST windows, so the worst case should be `start == end`. You should never encounter a `start > end` due to DST windows. Or are you talking about potentially pre-existing tests that rely on undefined behaviour by passing in `start` values that are after `end` values and it used to "just work"? This should not be an issue because `ComputeRealRange` swaps out-of-order `start` and `end` before beginning its computations.

File: Source/Bogus/DataSets/Date.cs (hunk 5 above)
Comment (logiclrd) [reply Source/Bogus/DataSets/Date.cs:134]: This is a good point, functions that call `SystemClock` multiple times should only call it once and use it as a consistent reference. Otherwise, in extremely rare cases, two consecutive calls will return different values, and the second one _could_ be earlier than the first one. I will make this change.

File: Source/Bogus/DataSets/Date.cs (hunk 6 above)
Comment (logiclrd) [reply Source/Bogus/DataSets/Date.cs:270]: I think the visible representation of the branching is a feature, not a bug, personally. :-P Reducing one level of nesting for 20 lines of code comes at a cost of the code not _looking_ like the branch that it is. Also, even though it is only a couple of lines of code in one statement, the `return` statement _is_ doing work and that would need to be duplicated.

File: Source/Bogus.Tests/DataSetTests/DateTest.cs (hunk 10 above)
Comment (logiclrd) [reply Source/Bogus.Tests/DataSetTests/DateTest.cs:11]: I am horrified by the idea of unit tests altering system-wide settings. :-P We are between a rock and a hard place on this one. In my opinion, only enabling the tests when the ambient environment supports them is the lesser evil.

File: Source/Bogus.Tests/DataSetTests/DateTest.cs (hunk 11 above)
Comment (logiclrd) [reply Source/Bogus.Tests/DataSetTests/DateTest.cs:395]: This sounds like a good thing but it's not something I'm super familiar with. Can you explain in more detail?

File: Source/Bogus.Tests/DataSetTests/DateTest.cs (hunk 12 above)
Comment (logiclrd) [reply Source/Bogus.Tests/DataSetTests/DateTest.cs:435]: I am in favour of this kind of thing :-)

File: Source/Bogus.Tests/DataSetTests/DateTest.cs (hunk 13 above)
Comment (logiclrd) [reply Source/Bogus.Tests/DataSetTests/DateTest.cs:600]: As far as I can tell, it is always correct _in .NET specifically_. Consider the following documentation, for instance:

> CalendarWeekRule Enum (System.Globalization)
> The first week based on the FirstDay value can have one to seven days. The first week based on the FirstFullWeek value **always has seven days**. The first week ...

([link](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.calendarweekrule?view=net-6.0))

...and the fact that the `DayOfWeek` enum has _exactly_ the western 7 days of the week in it.

File: Source/Bogus/DataSets/Date.cs (hunk 6 above)
Comment (ChristoWolf) [reply Source/Bogus/DataSets/Date.cs:270]: Not a bug of course, just a matter of preference :)

File: Source/Bogus.Tests/DataSetTests/DateTest.cs (hunk 11 above)
Comment (ChristoWolf) [reply Source/Bogus.Tests/DataSetTests/DateTest.cs:395]: [This](https://fluentassertions.com/introduction#assertion-scopes) should be good, let me know if I can give you more insight!

File: Source/Bogus.Tests/DataSetTests/DateTest.cs (hunk 13 above)
Comment (ChristoWolf) [reply Source/Bogus.Tests/DataSetTests/DateTest.cs:600]: Good to know, thanks!

File: Source/Bogus.Tests/DataSetTests/DateTest.cs (hunk 11 above)
Comment (logiclrd) [reply Source/Bogus.Tests/DataSetTests/DateTest.cs:395]: Ah, nice :-)

**Comment by logiclrd** [comment]
(Somehow I committed/pushed changes that just randomly deleted 7 characters from the middle of a line, force-push makes that commit into what it was supposed to be.

```
diff --git a/Source/Bogus.Tests/DataSetTests/DateTest.cs b/Source/Bogus.Tests/DataSetTests/DateTest.cs
index b8fbdc0..4dc3603 100644
--- a/Source/Bogus.Tests/DataSetTests/DateTest.cs
+++ b/Source/Bogus.Tests/DataSetTests/DateTest.cs
@@ -460,7 +460,7 @@ public void will_adjust_end_time_to_avoid_dst_transition()
          var transitionEndTime = transitionStartTime + effectiveRule.DaylightDelta;

          var windowStart = transitionStartTime.AddMinutes(-30);
-         var nd = transitionStartTime + TimeSpan.FromTicks((transitionEndTime - transitionStartTime).Ticks / 2);
+         var windowEnd = transitionStartTime + TimeSpan.FromTicks((transitionEndTime - transitionStartTime).Ticks / 2);

          // Act & Assert
          using (new AssertionScope())
```
)

**Comment by logiclrd** [comment]
> I would also prefer consolidation of the test initialization, what's your opinion @bchavez?

Making the tests DAMP rather than DRY is a deliberate choice here. Obviously, if those who own the solution have a strong preference then I'll defer to that, but it has repeatedly been my experience that factoring out parts of automated tests almost always results in the tests being harder to grok and harder to maintain.

File: Source/Bogus.Tests/DataSetTests/DateTest.cs (hunk 10 above)
Comment (logiclrd) [reply Source/Bogus.Tests/DataSetTests/DateTest.cs:12]: In my opinion, the correct solution to this is to have the test-running environment prepare virtual machines with desired configuration on which to run the tests. This, of course, doesn't solve the problem for running tests locally, but _that_ I feel _is_ a place where if a developer really needs it, they can temporarily (manually) change their time zone settings.

**Comment by bchavez** [comment]
Quick update:
Finally got the CI/CD build system working again. 😰 PR #471 is now merged.
Next, I'll be going through this PR and @ChristoWolf 's PR review notes as the next to-do item for Bogus soon.

Thanks for the patience.

**Comment by logiclrd** [comment]
How's it looking? :-)

**Comment by bchavez** [comment]
> How's it looking? :-)

Rebased with latest on master; let's see how that goes --
